Deploying_Seq2Seq_Model_with_Hybrid_Frontend.ipynb

```diff
@@ -0,0 +1,720 @@
+    "\n",
+    "### Hybrid Frontend Notes:\n",
+    "\n",
+    "Similarly to the `EncoderRNN`, this module does not contain any data-dependent control flow. Therefore, we can once again use **tracing** to convert this model to Torch Script after it is initialized and its parameters are loaded."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class LuongAttnDecoderRNN(nn.Module):\n",
+    "    def __init__(self, attn_model, embedding, hidden_size, output_size, n_layers=1, dropout=0.1):\n",
+    "        super(LuongAttnDecoderRNN, self).__init__()\n",
+    "\n",
+    "        # Keep for reference\n",
+    "        self.attn_model = attn_model\n",
+    "        self.hidden_size = hidden_size\n",
+    "        self.output_size = output_size\n",
+    "        self.n_layers = n_layers\n",
+    "        self.dropout = dropout\n",
+    "\n",
+    "        # Define layers\n",
+    "        self.embedding = embedding\n",
+    "        self.embedding_dropout = nn.Dropout(dropout)\n",
+    "        self.gru = nn.GRU(hidden_size, hidden_size, n_layers, dropout=(0 if n_layers == 1 else dropout))\n",
+    "        self.concat = nn.Linear(hidden_size * 2, hidden_size)\n",
+    "        self.out = nn.Linear(hidden_size, output_size)\n",
+    "\n",
+    "        self.attn = Attn(attn_model, hidden_size)\n",
+    "\n",
+    "    def forward(self, input_seq, last_hidden, encoder_outputs):\n",
+    "        # Note: we run this one step (word) at a time\n",
+    "        # Get embedding of current input word\n",
+    "        embedded = self.embedding(input_seq)\n",
+    "        embedded = self.embedding_dropout(embedded)\n",
+    "        # Forward through unidirectional GRU\n",
+    "        rnn_output, hidden = self.gru(embedded, last_hidden)\n",
+    "        # Calculate attention weights from the current GRU output\n",
+    "        attn_weights = self.attn(rnn_output, encoder_outputs)\n",
+    "        # Multiply attention weights to encoder outputs to get new \"weighted sum\" context vector\n",
+    "        context = attn_weights.bmm(encoder_outputs.transpose(0, 1))\n",
+    "        # Concatenate weighted context vector and GRU output using Luong eq. 5\n",
+    "        rnn_output = rnn_output.squeeze(0)\n",
+    "        context = context.squeeze(1)\n",
+    "        concat_input = torch.cat((rnn_output, context), 1)\n",
+    "        concat_output = torch.tanh(self.concat(concat_input))\n",
+    "        # Predict next word using Luong eq. 6\n",
+    "        output = self.out(concat_output)\n",
+    "        output = F.softmax(output, dim=1)\n",
+    "        # Return output and final hidden state\n",
+    "        return output, hidden"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data Handling\n",
```

Comment:
Feels weird having this here and not earlier on. I don't think it depends on anything that's been covered so far, and understanding what all this "embedding" stuff is doing requires understanding how we're representing our sequences.